Feedback 3 (#34)
mimuw-jnp2-rust/mimuw-jnp2-rust.github.io: content/lessons/11_feedback3/index.md

use the `fuse()` method.

### or_else

```rust
impl<I, J> Iterator for InterleaveIterator<I, J>
// where etc.
{
    fn next(&mut self) -> Option<Self::Item> {
        let mut ret_val;

        match self.next_from_a {
            true => {
                ret_val = self.i.next();
                if ret_val.is_none() {
                    ret_val = self.j.next()
                }
            }
            false => {
                ret_val = self.j.next();
                if ret_val.is_none() {
                    ret_val = self.i.next()
                }
            }
        }

        self.next_from_i = !self.next_from_i;

        ret_val
    }
}
```

Even though in this definition we don't have the excessive `bool`s,
it can still be written a lot more concisely using `or_else`.

```rust
impl<I, J> Iterator for InterleaveIterator<I, J>
// where etc.
{
    fn next(&mut self) -> Option<Self::Item> {
        self.next_from_i = !self.next_from_i;
        if self.first {
            self.i.next().or_else(|| self.j.next())
        } else {
            self.j.next().or_else(|| self.i.next())
        }
    }
}
```

### Why not `or`?

> The `or` method evaluates the argument event if it's not used.

Reviewer: ```suggestion
The `or` method evaluates the argument even if it's not used (eager evaluation).
```

        let mut counter = 0;
        let shreds = self
            .shreds
            .into_iter()
            .fold(vec![], |mut acc, el| {
                acc.push((el, acc.len()));
                acc
            })
            .filter(|(_, nr)| nr % rhs == 0)
            .map(|(el, _)| el)
            .collect();
        Shreds { shreds }
    }
}
```

Instead of `fold` we can use `enumerate` to pair each element with its index.

```rust
impl<'a> Div<usize> for Shreds<'a> {
    type Output = Shreds<'a>;

    fn div(self, rhs: usize) -> Self::Output {
        let mut counter = 0;
        let shreds = self
            .shreds
            .into_iter()
            .enumerate()
            .filter(|(_, nr)| nr % rhs == 0)
            .map(|(el, _)| el)
            .collect();
        Shreds { shreds }
    }
}
```

However, it can be simplified even more. What we're doing here is basically reimplementing
`step_by` :)

```rust
impl<'a> Div<usize> for Shreds<'a> {
    type Output = Shreds<'a>;
    fn div(self, rhs: usize) -> Self::Output {
        Shreds {
            shreds: self.shreds.into_iter().step_by(rhs).collect(),
        }
    }
}
```

> 
> ```rust
> impl<'a> Div<usize> for Shreds<'a> {
>     type Output = Shreds<'a>;
> 
>     fn div(self, rhs: usize) -> Self::Output {
>         let mut counter = 0;
>         let shreds = self
>             .shreds
>             .into_iter()
>             .enumerate()
>             .filter_map(|(_, nr)| nr % rhs == 0)
>             .map(|(el, _)| el)
>             .collect();
>         Shreds { shreds }
>     }
> }
> ```

Reviewer: ```suggestion
```

Reviewer: This should not be here, right?

### What's `collect()`?

It's not magic. We can collect the elements of an iterator into any type which implements
the appropriate `FromIterator` [trait](https://doc.rust-lang.org/std/iter/trait.FromIterator.html).

## Shredding usize

Instead of

```rust
impl Shredder for usize {
    fn shred(&self) -> Shreds {
        let mut elements = Vec::new();
        for i in self.to_string().chars() {
            let dig = i as usize - '0' as usize;
            let val_dig = crate::value::Digit::new(dig);
            elements.push(Value::Digit(val_dig));
        }
        Shreds { elements }
    }
}
```

it's better to use the modulo operator and division to get the digits. Why? Converting a number to
string requires an additional heap allocation.

## Make illegal states unrepresentable

Some people used an i8 or some other integer type to keep track of whose turn it is. But the only

> values that were only ever used were 0 and 1. It means that there was a lot of cases where the

Reviewer: ```suggestion
values that were ever used were 0 and 1. It means that there was a lot of cases where the
```
